doc/source/ray-overview/examples/web-crawler.ipynb

```diff
@@ -0,0 +1,241 @@
+    "`pip install beautifulsoup4 ray`.\n",
+    "\n",
+    "First, we'll define a function called `find_links` which takes a starting page (`start_url`) to crawl,\n",
+    "and we'll take the Ray documentation as example of such a starting point.\n",
+    "Our crawler simply extracts all available links from the starting URL that contain a given `base_url`\n",
+    "(e.g. in our example we only want to follow links on `http://docs.ray.io`, not any external links).\n",
+    "The `find_links` function is then called recursively with all the links we found this way, until a\n",
+    "certain depth is reached.\n",
+    "\n",
+    "To extract the links from HTML elements on a site, we define a little helper function called\n",
+    "`extract_links`, which takes care of handling relative URLs properly and sets a limit on the\n",
+    "number of links returned from a site (`max_results`) to control the runtime of the crawler more easily.\n",
+    "\n",
+    "Here's the full implementation:"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 154,
+   "outputs": [],
+   "source": [
+    "import requests\n",
+    "from bs4 import BeautifulSoup\n",
+    "\n",
+    "def extract_links(elements, base_url, max_results=100):\n",
+    "    links = []\n",
+    "    for e in elements:\n",
+    "        url = e[\"href\"]\n",
+    "        if \"https://\" not in url:\n",
+    "            url = base_url + url\n",
+    "        if base_url in url:\n",
+    "            links.append(url)\n",
+    "    return set(links[:max_results])\n",
+    "\n",
+    "\n",
+    "def find_links(start_url, base_url, depth=2):\n",
+    "    if depth == 0:\n",
+    "        return set()\n",
+    "\n",
+    "    page = requests.get(start_url)\n",
+    "    soup = BeautifulSoup(page.content, \"html.parser\")\n",
+    "    elements = soup.find_all(\"a\", href=True)\n",
+    "    links = extract_links(elements, base_url)\n",
+    "\n",
+    "    for url in links:\n",
+    "        new_links = find_links(url, base_url, depth-1)\n",
+    "        links = links.union(new_links)\n",
+    "    return links"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "Let's define a starting and base URL and crawl the Ray docs to a `depth` of 2."
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 162,
+   "outputs": [],
+   "source": [
+    "base = \"https://docs.ray.io/en/latest/\"\n",
+    "docs = base + \"index.html\""
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 163,
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 19.3 s, sys: 340 ms, total: 19.7 s\n",
+      "Wall time: 25.8 s\n"
+     ]
+    },
+    {
```

Comment:
somehow when this renders, it looks like code, so it'd be nice if we can differentiate somehow...

Reply:
yes, good point. I made the output bold for now. We can iterate on that.